### ISSUE TITLE
2023-06->2023-09 Seems to have broke dependency graph management in our project 

### ISSUE THREAD

**Issue by mebigfatguy** [eclipse-jdt/eclipse.jdt.core#1654]
We have a large jdk8 project that includes over 35 projects in one eclipse workspace. In 2023.06, eclipse could handle this with ease finding and building dependent classes between projects.

Upgrading to 2023.09, the eclipse build is in a continual  state of broken with errors of the form

The project was not built since its build path is incomplete. Cannot find the class file for com.acme.common.entity.Activity. 

Fix the build path then try building this project	common		

Unknown	Java Problem

--

No  amount of refreshing, cleaning, building from the command line, etc, will fix  it. If you roll back  to  2023.06 it works fine. There are no circular dependencies between projects. The eclipse logs do not seem to have anything relevant.

Happens for all team members whether on linux or windows

Version: 2023-09 (4.29.0)
Build id: 20230907-1323

the below, works fine

https://eclipse.org/downloads/packages/release/2023-06/r

**Comment by mebigfatguy** [comment]
@jukzi no annotation processing in eclipse.

**Comment by haubi** [comment]
> > So I have been able to bisect this problem
> 
> Excellent, many thanks! @haubi: Any errors in the log? Any chance to get a reduced self contained example project/set of projects that show the problem?

Reducing some self contained example would be a real challenge for me, since on one hand the application consists of 500+ projects and 80k+ files, and on the other hand I'm not familiar yet with creating such examples in general.

There's nothing on the console, and there's this one message in `.metadata/.log`:
```
!ENTRY org.eclipse.jdt.apt.pluggable.core 1 1 2023-11-30 12:09:46.709
!MESSAGE Hibernate JPA 2 Static-Metamodel Generator 5.6.15.Final
```
Anything you can recommend for enabling some logs?

The only sign of a problem is this set of Java compilation errors in some `UsingAnnotatedType.java` (where `AnnotatedType` and it's superclasses use some annotations from `javax.persistence`, successfully processed by the `org.hibernate:hibernate-jpamodelgen:5.6.15.Final` maven artifact):
- `AnnotatedType_ cannot be resolved` 
- `AnnotatedType_ cannot be resolved to a variable`
- `The import package.of.AnnotatedType_ cannot be resolved`

After some no-op change (add some irrelevant blank) on the `UsingAnnotatedType.java` files, one by one, the incremental build succeeds.
But the build after cleaning that project (=full-build?) always fails, which is the same as any initial build I guess.

The more I play around (using the Eclipse 2023-09 I did above git bisect with), the more I'm convinced this is some synchronization issue somewhere, because there is one workflow I'm able to reliably get a successful build for this particular project (not useable as general workaround, though):

1. Remove all projects (500+ here) from the workspace that would be compiled after the first failing project
2. Prepare some Git-Stash that removes the javax.persistence annotations from those `AnnotatedType.java` files, whose generated `AnnotatedType_` erroneously `cannot be resolved` during Clean (=Full) Build
3. Keep Progress View visible, with `Show sleeping and system operations` enabled
4. Have Auto-Build enabled
5. Have the Source unmodified (annotations intact)
6. Clean the project: the Auto-Build ends up with above `AnnotatedType_ cannot be resolved` errors
7. `Apply Stashed Changes` to remove annotations from the `AnnotatedType.java` files
8. Let the Auto-Build finish, will end up with more errors (which is expected)
9. Disable Auto-Build (**<-- this one makes a difference here**)
10. In Git Staging View, `Replace with HEAD Revision` on the `AnnotatedType.java` files that got the annotations dropped in 7.
11. Wait until Progress View is silent
12. Enable Auto-Build
13. **All** the errors are gone

Sometimes disabling Auto-Build (9.) is not necessary, but each time I do, I get a successful build.
Eventually, it is some Indexer Jobs that need to finish before the Auto-Build can be successful in this situation - which clearly is not true for an initial or even headless build.

I'm adding all this in the hope to ring someone's bell regarding synchronization issues, maybe even related to https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1367.

**Comment by haubi** [comment]
FWIW: Confirming same behaviour using current Eclipse 2023-12, and even with current master of jdt.core per https://github.com/eclipse-jdt/eclipse.jdt.core/commit/b2cad64dfc0086a314e8b5d842a4644aaede021a.

**Comment by haubi** [comment]
Since I can reproduce it using the application here, I spent some time debugging JDT:
So far, the situation triggering the full build to fail seems to be when some class having annotations to process, does import another class that is being created by some annotation processor - eventually in a different package only, but in the same project.
But I''ve no idea yet how this situation is supposed to be handled by JDT.

**Comment by haubi** [comment]
Yep, I can reproduce the issue in a small test case now.

The important bits for a class to break JDT during Full Build, while the Incremental Build succeeds:
- **import** some class generated by some annotation processor
- use some other annotation, passing some **parameter value** that is **defined locally** within the same project

So I have set up a jdt-master workspace (using [JDTConfiguration.setup](https://raw.githubusercontent.com/eclipse-jdt/eclipse.jdt/master/org.eclipse.jdt.releng/JDTConfiguration.setup)), but looking at the existing tests, I seem to be lost on how and where best to add this new test case.
Please advice on which existing test I could look at to get some ideas.

What I want to add for the test case (please advice on how to reduce the Maven Artifacts even more, if sensible):
1. Maven Artifact `org.hibernate:hibernate-jpamodelgen:5.6.15.Final`:
Registered (via .factorypath) as the Annotation Processor
2. Maven Artifact `jakarta.xml.bind:jakarta.xml.bind-api:2.3.3`:
Registered (via .factorypath), required by (above version of) hibernate-jpamodelgen
3. Maven Artifact `javax.persistence:javax.persistence-api:2.2`:
Provides the annotations (above version of) hibernate-jpamodelgen is acting on
4. Class file `base/SomeEntity.java`:
```
package base;
import javax.persistence.Entity;
@Entity
public class SomeEntity {
}
```
5. Class file `base/LocalAnnotation.java`:
```
package base;
public @interface LocalAnnotation {
    public static final String LocalValue = "LocalValue";
    String value() default "";
}
```
6. Class file `combined/Combined.java`:
```
package combined;
import base.LocalAnnotation;
import base.SomeEntity_;
@LocalAnnotation(value = LocalAnnotation.LocalValue)
public class Combined {
    public SomeEntity_ someEntity;
}
```
7. Eclipse Project settings as necessary

**Comment by jukzi** [comment]
@haubi there was a testcase submitted together with "maxCompiledUnitsAtOnce":
org.eclipse.jdt.core.tests.builder.BasicBuildTests.testbBug386901()
can you adapt a copy of that to your example, please?

**Comment by haubi** [comment]
Since I may want to use `AptConfig.setFactoryPath`, this new test should go into `org.eclipse.jdt.apt.tests`, otherwise I seem to get a dependency circle, no?

**Comment by haubi** [comment]
@iloveeclipse A test case is available in PR #1698 now, created with the idea of becoming submitted to JDT.Core.

@mebigfatguy In my situation, in those class files that JDT fails to compile now, the workaround is:
- For those classes JDT fails to resolve, use their fully qualified name instead of importing them.

Does this workaround help in your situation as well?

**Comment by mebigfatguy** [comment]
No, i suspect my issue is different.

problem is as simple as trying to import a class in the same package into
another class -- fails to find it. No annotation processing.

Description Resource Path Location Type
The project was not built since its build path is incomplete. Cannot find
the class file for com.acme.pgbu.common.entity.Activity. Fix the build path
then try building this project api Unknown Java Problem

public class GqlActivityRiskDAO extends AbstractGqlDAO<ActivityRisk> {

The type com.acme.pgbu.common.entity.Activity cannot be resolved. It is
indirectly referenced from required type

com.acme.pgbu.common.entity.ActivityRisk

**Comment by stephan-herrmann** [comment]
I'm preparing a fix that avoids "The import p1.gen cannot be resolved to a type" as raised by #1698 

I'm not sure how this matches the above comment "i don't have an import statement of course". As my fix is specific to imports we may be facing two different issues after all.

**Comment by mebigfatguy** [comment]
Ok i've figured out the problem in my program here goes (i have annotations but am doing no annotation processing)

```
package eclipse1654.p1;

import eclipse1654.p2.Inf;

@Anno(Cls.CON)
public class Cls implements Inf {
}
```

```
package eclipse1654.p1;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface Anno {

    String value();
}
```

```
package eclipse1654.p2;

public interface Inf {

    String CON = "Con";
}
```

building class Cls from the command line works fine. Eclipse shows error on the Cls.CON in the 
@Anno(Cls.CON)

If instead you use

@Anno(Inf.CON)

it works fine

**Comment by jukzi** [comment]
@mebigfatguy mind to contribute a Junit test that shows the error? that will make it more likely to be fixed. Also consider to open a new dedicated issue to separate the superclass-issue from the annotation-processing-issue also mentioned in the thread

**Comment by guw** [comment]
Confirming that the issues is related to a problem I noticed (and fixed in ECJ 4.31RC1)
https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/2076

**Comment by stephan-herrmann** [comment]
> So can we close this ticket now?

The issue observed by @haubi has been fixed via #1698

As for the issues observed by @mebigfatguy  and @guw I suggest to wait until master has been re-opened and #2059 is merged. At that point we can close this issue, OK?

**Comment by stephan-herrmann** [comment]
Closing as per previous comment.
